File: draft_design_pattern/flyweight.nim
import tables
import strformat


type
  KarakTea = ref object
    val: string

  TeaMaker = ref object
    availableTea: Table[string, KarakTea]

  TeaShop = ref object
    orders: OrderedTable[int, KarakTea]
    teaMaker: TeaMaker


proc make(self: TeaMaker, preference: string): KarakTea =
  if not self.availableTea.hasKey(preference):
    self.availableTea[preference] = KarakTea(val: preference)
  return self.availableTea[preference]


proc takeOrder(self: TeaShop, teaType: string, table: int) =
  self.orders[table] = self.teaMaker.make(teaType)

proc serve(self: TeaShop) =
  for key, value in self.orders:
    echo fmt"テーブル# {key}にお茶を出す {value.val} {repr(value)}"


proc main() =
  let teaMaker: TeaMaker = TeaMaker()
  let shop: TeaShop = TeaShop(teaMaker: teaMaker)

  shop.takeOrder("砂糖少なめ", 1)
  shop.takeOrder("ミルク多め", 2)
  shop.takeOrder("砂糖なし", 5)
  shop.takeOrder("砂糖なし", 6)
  shop.takeOrder("砂糖少なめ", 7)

  shop.serve()


main()

File: draft_design_pattern/flyweight.ts
export {};

let objectID = 1;

class KarakTea {
  public objectID = objectID++;
  constructor(public val: string) {}
}

class TeaMaker {
  private availableTea: { [key: string]: KarakTea } = {};
  constructor() {}
  make(preference: string): KarakTea {
    if (!this.availableTea[preference]) {
      this.availableTea[preference] = new KarakTea(preference);
    }
    return this.availableTea[preference];
  }
}

class TeaShop {
  private orders: { [key: number]: KarakTea } = {};
  private teaMaker: TeaMaker;
  constructor(teaMaker: TeaMaker) {
    this.teaMaker = teaMaker;
  }
  takeOrder(teaType: string, table: number) {
    this.orders[table] = this.teaMaker.make(teaType);
  }
  serve() {
    for (const k in this.orders) {
      console.log(
        `テーブル# ${k}にお茶を出す ${this.orders[k].val} , objectID = ${
          this.orders[k].objectID
        }`,
      );
    }
  }
}

const teaMaker = new TeaMaker();
const shop = new TeaShop(teaMaker);

shop.takeOrder("砂糖少なめ", 1);
shop.takeOrder("ミルク多め", 2);
shop.takeOrder("砂糖なし", 5);
shop.takeOrder("砂糖なし", 6);
shop.takeOrder("砂糖少なめ", 7);

shop.serve();
